src/main/java/javaslang/control/Option.java

```diff
@@ -46,6 +46,22 @@
     }
 
     /**
+     * Reduces many {@code Option}s into a single {@code Option} by transforming an
+     * {@code Iterable<Option<? extends T>>} into a {@code Option<Seq<T>>}. If any of
+     * the Options are {@link Option.None}, then this returns {@link Option.None}.
+     *
+     * @param values An {@code Iterable} of {@code Option}s.
+     * @param <T> type of the Options.
+     * @return An {@code Option} of a {@link Seq} of results.
+     */
+    static <T> Option<Seq<T>> sequence(Iterable<? extends Option<? extends T>> values) {
+        final Option<Seq<T>> zero = Option.of(List.empty());
+        final BiFunction<Option<Seq<T>>, Option<? extends T>, Option<Seq<T>>> f =
+                (result, option) -> result.flatMap(sequence -> option.map(sequence::append));
+        return Iterator.ofAll(values).foldLeft(zero, f);
```

Comment:
foldLeft is the most concise solution here. But we cannot shortcut the loop if we visit the first None (see #779).

I would suggest to optimize the internal implementation similar to this:

```java
List<T> list = List.empty();
for (T value : values) {
    if (value.isEmpty) {
        return Option.none();
    }
    list = list.prepend(value.get());
}
return Option.some(list.reverse());
```

I know, from a functional perspective this code looks ugly. But it does perform better on the non-happy path.